common/src/main/java/com/microsoft/identity/common/internal/controllers/ExceptionAdapter.java

```diff
@@ -137,111 +130,131 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
                 );
             }
 
-            if (null == outErr) {
-                outErr = new ServiceException(
+            outErr = getExceptionFromOAuthError(
                         tokenErrorResponse.getError(),
+                        tokenErrorResponse.getSubError(),
                         tokenErrorResponse.getErrorDescription(),
-                        null
-                );
-            }
+                    synthesizeHttpResponse(tokenErrorResponse.getStatusCode(),
+                            tokenErrorResponse.getResponseHeadersJson(),
+                            tokenErrorResponse.getResponseBody()));
 
-            applyCliTelemInfo(tokenResult, outErr);
+            applyCliTelemInfo(tokenResult.getCliTelemInfo(), outErr);
+        }
 
-            applyHttpErrorResponseData(
-                    outErr,
-                    tokenErrorResponse.getStatusCode(),
-                    tokenErrorResponse.getResponseHeadersJson(),
-                    tokenErrorResponse.getResponseBody()
-            );
+        return null;
+    }
 
-            return outErr;
+    /**
+     * Determine if an exception owning the given error codes should be converted into UiRequiredException.
+     *
+     * @param oAuthError
+     * @param oAuthSubError
+     * @return boolean
+     * */
+    private static boolean shouldBeConvertedToUiRequiredException(final String oAuthError,
+                                                                  final String oAuthSubError){
+        // Invalid_grant doesn't necessarily requires UI protocol-wise.
+        // We simplify our logic because this layer is also used by MSAL.
+        if (oAuthError.equalsIgnoreCase(AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT)){
```

Comment:
Is this where we plan to add invalid_resource for now?

Reply:
I'm not sure what that is. This is only for converting Exception type. 

For error-handling (inside Broker), we should still use the error code + subError code.